packages/salesforce-adapter/src/custom_references/profile_permission_set_utils.ts

```diff
@@ -0,0 +1,350 @@
+ * limitations under the License.
+ */
+import _ from 'lodash'
+import { collections, promises } from '@salto-io/lowerdash'
+import {
+  ChangeError,
+  ElemID,
+  FixElementsFunc,
+  InstanceElement,
+  ReadOnlyElementsSource,
+  Values,
+} from '@salto-io/adapter-api'
+import { logger } from '@salto-io/logging'
+import { Types } from '../transformers/transformer'
+import {
+  APEX_CLASS_METADATA_TYPE,
+  APEX_PAGE_METADATA_TYPE,
+  API_NAME_SEPARATOR,
+  CUSTOM_APPLICATION_METADATA_TYPE,
+  FLOW_METADATA_TYPE,
+  LAYOUT_TYPE_ID_METADATA_TYPE,
+  RECORD_TYPE_METADATA_TYPE,
+  SALESFORCE,
+} from '../constants'
+
+const { makeArray } = collections.array
+const { awu } = collections.asynciterable
+const { pickAsync } = promises.object
+
+const log = logger(module)
+
+enum section {
+  APEX_CLASS = 'classAccesses',
+  APEX_PAGE = 'pageAccesses',
+  APP_VISIBILITY = 'applicationVisibilities',
+  FIELD_PERMISSIONS = 'fieldPermissions',
+  FLOW = 'flowAccesses',
+  LAYOUTS = 'layoutAssignments',
+  OBJECT = 'objectPermissions',
+  RECORD_TYPE = 'recordTypeVisibilities',
+}
+
+const FIELD_NO_ACCESS = 'NoAccess'
+
+const getMetadataElementName = (fullName: string): string =>
+  Types.getElemId(fullName.replace(API_NAME_SEPARATOR, '_'), true).name
+
+type ReferenceInSection = {
+  sourceField?: string
+  target: ElemID
+}
+
+type RefTargetsGetter = (
+  sectionEntry: Values,
+  sectionEntryKey: string,
+) => ReferenceInSection[]
+
+type ReferenceFromSectionParams = {
+  filter?: (sectionEntry: Values) => boolean
+  targetsGetter: RefTargetsGetter
+}
+
+const mapSectionEntries = <T>(
+  profile: InstanceElement,
+  sectionName: section,
+  { filter = () => true, targetsGetter }: ReferenceFromSectionParams,
+  f: (sectionEntryKey: string, target: ElemID, sourceField?: string) => T,
+): T[] => {
+  const sectionValue = profile.value[sectionName]
+  if (!_.isPlainObject(sectionValue)) {
+    if (sectionValue !== undefined) {
+      log.warn(
+        'Section %s of %s is not an object, skipping.',
+        sectionName,
+        profile.elemID,
+      )
+    }
+    return []
+  }
+  return Object.entries(sectionValue as Values)
+    .filter(([, sectionEntry]) => filter(sectionEntry))
+    .flatMap(([sectionEntryKey, sectionEntry]) => {
+      const targets = targetsGetter(sectionEntry, sectionEntryKey)
+      return targets.map(({ target, sourceField }) =>
+        f(sectionEntryKey, target, sourceField),
+      )
+    })
+}
+
+const isEnabled = (sectionEntry: Values): boolean =>
+  sectionEntry.enabled === true
+
+const isAnyAccessEnabledForObject = (
+  objectAccessSectionEntry: Values,
+): boolean =>
+  [
+    objectAccessSectionEntry.allowCreate,
+    objectAccessSectionEntry.allowDelete,
+    objectAccessSectionEntry.allowEdit,
+    objectAccessSectionEntry.allowRead,
+    objectAccessSectionEntry.modifyAllRecords,
+    objectAccessSectionEntry.viewAllRecords,
+  ].some((permission) => permission === true)
```

Comment:
Not sure what is considered better style, but this could just use the identity here:
```suggestion
  ].some((x) => x)
```
Actually it's a bit weird to me that you can't call `some` with no argument and get this behaviour, but that's JS for you. Maybe I'll write a proposal.

Reply:
For my part I'd rather be explicit about this sort of thing. I wonder about the semantic difference, though - what do we want to happen if there's a bug and any of these fields contain something that's not a boolean?